lib/auth/saml.go

```diff
@@ -336,87 +415,107 @@ func (a *Server) ValidateSAMLResponse(samlResponse string) (*SAMLAuthResponse, e
 		}
 		return nil, trace.Wrap(err)
 	}
+
 	event.Status.Success = true
-	event.User = re.auth.Username
+	event.User = auth.Username
 	event.Code = events.UserSSOLoginCode
+	if diagCtx.testFlow {
+		event.Code = events.UserSSOTestFlowLoginCode
+	}
 
 	if err := a.emitter.EmitAuditEvent(a.closeCtx, event); err != nil {
 		log.WithError(err).Warn("Failed to emit SAML login event.")
 	}
 
-	return &re.auth, nil
-}
-
-type samlAuthResponse struct {
-	auth                SAMLAuthResponse
-	attributeStatements map[string][]string
-}
-
-func (a *Server) validateSAMLResponse(samlResponse string) (*samlAuthResponse, error) {
-	ctx := context.TODO()
-	requestID, err := parseSAMLInResponseTo(samlResponse)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	request, err := a.Identity.GetSAMLAuthRequest(requestID)
+	return auth, nil
+}
+
+func (a *Server) validateSAMLResponse(ctx context.Context, diagCtx *ssoDiagContext, samlResponse string) (*SAMLAuthResponse, error) {
+	requestID, err := ParseSAMLInResponseTo(samlResponse)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	connector, err := a.Identity.GetSAMLConnector(ctx, request.ConnectorID, true)
+	diagCtx.requestID = requestID
+
+	request, err := a.Identity.GetSAMLAuthRequest(ctx, requestID)
 	if err != nil {
-		return nil, trace.Wrap(err)
+		return nil, trace.Wrap(err).AddUserMessage("Failed to get SAML Auth Request")
```

Comment:
Is there a difference between `trace.Wrap(err).AddUserMessage("xxx")` and `trace.Wrap(err, "xxx")`?

Reply:
There is tiny, but it is in favor of using `trace.Wrap(err, "xxx")`. I'll convert the code to use that.